### ISSUE TITLE
Estimating memory requirements for arks-long (core dump)

### ISSUE THREAD

**Issue by Nicholas-Kron** [bcgsc/LongStitch#67]
I am attempting to scaffold a 2.2 GB genome using 194GB of hifi reads (uncompressed fastq). I have increased the memory allocation a few times but every time I run the pipeline I get a core dump. The pipeline gets through 5 rounds of ntLinks but fails at the ```# Run arcs-long``` step of the ```make``` due to memory overdraw. My most recent run I gave it 16 x 3.5GB= 56GB of memory.  I haven't run into any issues with similar allocation for other scaffolders I've tested. Is there a  way to estimate how much memory I should request from my cluster? I can go as high as 250GB.

The command
```
longstitch ntLink-arks \
reads=hifi_reads \
draft=obeta_primary_assembly \
rounds=5
```

last successful line from output log:
```
LINKS -f obeta_primary_assembly.fa.k32.w100.z1000.ntLink.scaffolds.renamed.fa -s empty.fof -b obeta_primary_assembly.fa.k32.w100.z1000.ntLink.scaffolds_c4_m8-10000_cut250_k20_r0.05_e30000_z1000_l4_a0.3 -l 4 -a 0.3 -z 1000
```

Error log:
```
long-to-linked-pe v1.0: Using more than 6 threads does not scale, reverting to 6.
long-to-linked-pe v1.0: Using more than 6 threads does not scale, reverting to 6.
Number of contigs:582
Size of Contig Array:1165
File /dev/stdin opened.
time user=1232.27s system=714.03s elapsed=14602.29s cpu=13% memory=4 job= -l 250 -t 8 -m2000 hifi_reads.fq
time user=115239.20s system=643.36s elapsed=14625.24s cpu=792% memory=2820 job=
make[1]: *** [/nethome/n.kron/mambaforge/envs/longstitch/bin/share/arcs-1.2.5-0/bin/arcs-make:344: obeta_primary_assembly.fa.k32.w100.z1000.ntLink.scaffolds_c4_m8-10000_cut250_k20_r0.05_e30000_z1000_l4_a0.3.scaffolds.fa] Segmentation fault (core dumped)
make: *** [/nethome/n.kron/mambaforge/envs/longstitch/bin/share/longstitch-1.0.4-0/longstitch:301: obeta_primary_assembly.fa.k32.w100.z1000.ntLink.scaffolds_c4_m8-10000_cut250_k20_r0.05_e30000_z1000_l4_a0.3.scaffolds.fa] Error 2
```

**Comment by Nicholas-Kron** [comment]
Hi Lauren,

Thank you for your response. I figured something fishy was going on. To answer your questions:

1.  Below is the LSF out resource usage summary. Clearly the total allocation isn't even close to being taken advantage of. Perhaps a subprocess is only using a single core and drawing too much memory from that one core? The current allocation is 16 cores with 7.5GB per core. 

```
Exited with exit code 2.

Resource usage summary:

    CPU time :                                   167975.80 sec.
[longstitch_LSF_log.txt](https://github.com/bcgsc/LongStitch/files/11508639/longstitch_LSF_log.txt)

    Max Memory :                                 8984 MB
    Average Memory :                             2046.49 MB
    Total Requested Memory :                     120000.00 MB
[longstitch_env_list.txt](https://github.com/bcgsc/LongStitch/files/11508638/longstitch_env_list.txt)

    Delta Memory :                               111016.00 MB
    Max Processes :                              16
    Max Threads :                                62
```

2. I downloaded longstitch conta package via mamba. I have attached the env list.
[longstitch_env_list.txt](https://github.com/bcgsc/LongStitch/files/11508664/longstitch_env_list.txt)
3. Yes the demo ran through without any issue. 
4. Absolutely! No problem. I have attached the outfield log and the error log. 
[longstitch_LSF_log.txt](https://github.com/bcgsc/LongStitch/files/11508662/longstitch_LSF_log.txt)
[longstitch_LSF_err.txt](https://github.com/bcgsc/LongStitch/files/11508680/longstitch_LSF_err.txt)

Thank you for your assistance!

Nick

**Comment by lcoombe** [comment]
Thanks for all this information, Nick!

Ok, so the environment looks fine to me - I was wondering if there was something funny with the version of LINKS you were using, but it is the same as I use.

The most puzzling thing is that the `LINKS` command appears to be the one seg faulting, but it is just a perl script. I ran the command itself on my end, and for a typical human-sized assembly, it only required ~4GB of RAM. Also, it prints out a log message with the specified parameters before reading in any files, which I don't see in either of the logs that you provided. That suggests to me that there could be an issue with your `LINKS` installation, although the inconsistent piece is that the demo ran without issue.

Would you be able to do a couple of additional sanity checks with `LINKS`?
1. Can you run the command `LINKS` on the command-line by itself, and check that you see the help page printed out as expected, that it is the version expected based on the conda install, and the path corresponds to the conda install as well?
2. Do you have a file in your working directory with a suffix like: `*k20_r0.05_e30000_z1000_l4_a0.3.log`? If so, what are the contents of that file?

**Comment by Nicholas-Kron** [comment]
Hi Lauren,

Thank you for your patience.

1)  It seems the LINKs installation is broken for some reason. Running LINKS in the command line results in a segmentation fault. The conda environment has LINKs listed as v. 1.8.7.  I completely uninstalled the conda environment and reinstalled, but the same error occurs. I colleague of mine has successfully installed and run the longstich conda environment on the same HPC cluster, so I am not quite sure what the source of the error could be. I will try a clean install of mamba forge to see if that will fix it. 

2) I do not have *k20_r0.05_e30000_z1000_l4_a0.3.log, so execution must've failed before that point.

**Comment by lcoombe** [comment]
Hi @Nicholas-Kron,

I'm glad that we at least sorted out the cause of the issue! If LINKS is segmentation faulting, it makes sense that you do not have that log file I was asking about.
Yes, my best advice would be to start with a completely new and clean conda environment, and try installing `longstitch` again there. It's possible that some existing packages in the conda environment or something like that caused the issue with LINKS. 
Also make sure that you are specifying `bioconda` and `conda-forge` in your install command:
```
conda install -c bioconda -c conda-forge longstitch
```
(You can substitute `conda install` for `mamba install` if you are using mamba)

**Comment by Nicholas-Kron** [comment]
Hi Lauren,

It looks like the problem may not be with mamba or LINKS after all. When attempting to run another perl program, LTR_Finder_parallel, I am encountering the same issue (ie core dump on program activation). So this seems to be a perl issue on our HPC cluster. I will work with our support staff to see if we can get that fixed and I will let you know if that resolves the LINKS issues. Thanks again for your help!

**Comment by lcoombe** [comment]
Ah ok - hopefully your support staff can help get that sorted out! That makes sense that only the LINKS step would fail (that's the only stage that uses perl in LongStitch). Good luck!

**Comment by Nicholas-Kron** [comment]
Hi Lauren,

Following up, after clean install of mamba and longstitch and messing around with perl modules on the cluster based on advice from my HPC staff I got LINKS to work. I just had to export PERL5LIB environmental variable to the perl in the longstitch conda env and everything works fine. This fixed the other perl issues as well. I guess LINKS was trying to source perl libs from the wrong location. Thanks again for your help! I look forward to scaffolding with longstich!
